src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || size < index) {
+            throw new IndexOutOfBoundsException("The index " + index + " is invalid "
+                    + "the actual size of array is " + size);
+        }
+        Node<T> newNode = new Node<>(value);
+        if (first == null) {
+            first = last = newNode;
+        } else if (index == 0) {
+            newNode.next = first;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            last = newNode;
+        } else if (index > 0 && size > index) {
+            Node<T> current = first;
+            int c = 0;
+            while (current != null && c != index) {
+                current = current.next;
+                c++;
+            }
+            current.prev.next = newNode;
+            newNode.prev = current.prev;
+            current.prev = newNode;
+            newNode.next = current;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Node<T> newNode;
+        for (T t : list) {
+            newNode = new Node<>(t);
+            add(newNode.item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> originalNode = getNodeByIndex(index);
+        T oldValue = originalNode.item;
+        originalNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement;
+        if (index == 0) {
+            removedElement = first.item;
+            first = first.next;
+            if (first == null) {
+                last = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.item;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                last = prev;
+            }
+        }
+        size--;
+        return removedElement;
```

Comment:
Create a separate method `unlink(Node node)`. The logic of unlinking a node is used here and should also be used in the `remove(T object)` method. This will avoid code duplication and adhere to the DRY principle.